influxdb3_write/src/write_buffer/mod.rs

```diff
@@ -727,6 +740,62 @@ pub(crate) struct TableBatchMap<'a> {
     pub(crate) table_batches: HashMap<String, TableBatch>,
 }
 
+/// The 32 byte SHA256 digest of the full tag set for a line of measurement data
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct SeriesId([u8; 32]);
+
+impl std::fmt::Display for SeriesId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", hex::encode(self.0))
+    }
+}
+
+/// Produce a [`SeriesId`] from a raw line of line protocol / parsed [`Series`]
+///
+/// If the tag set is sorted, we parse out the original tags from the raw line
+/// protocol, instead of formatting a new string.
+///
+/// If the tags are not sorted, then we need to sort them, and allocate a new
+/// string to format them in the original 'key=val,key=val` form.
+///
+/// Generally, InfluxDB clients should be sending tags sorted.
+fn hash_series_id(raw_line: &str, series: &mut Series) -> SeriesId {
+    let bytes = if let Some(tag_set) = &mut series.tag_set {
+        if is_sorted(tag_set) {
+            // extract the raw string; if we are here, then the following `expect`s should
+            // never panic, there should be a tag set in the raw line protocol
```

Comment:
Is this because we expect this to be a valid lp line?

Reply:
Yeah, at this point the `tag_set` is a `TagSet`, which is parsed from a valid line of LP, i.e., the `raw_line` that we pass forward here. So, if it exists, then we should be able to do the following extraction to generate the hash without allocating a new string.